### PR TITLE
[CI] Run hardhat tests as a PR check

### LINKED ISSUE #10854: Running Buidler's test suite in our CI
From https://github.com/ethereum/solidity/issues/9237#issuecomment-678411675:

> This is now ready on our side. All you need to do to run our test suite with a custom compiler is:
> 
> ```shell
> git clone git@github.com:nomiclabs/hardhat.git
> cd hardhat
> git checkout development
> HARDHAT_TESTS_SOLC_VERSION="0.7.0" HARDHAT_TESTS_SOLC_PATH="/some/path/soljson-v0.7.0+commit.9e61f92b.js" scripts/run-tests-with-custom-solc.sh
> ```
> 
> (Note the version and path in the last command)
> 
> This will install all the necessary dependencies and run the relevant tests. The only requirement is node.js. Any of the currently actively maintained versions would do it.
> 
> If the tests fail, the script will exit with a non-zero status number.

**EDIT**: I slightly modified the quote above to use the updated snippet from https://github.com/ethereum/solidity/issues/9237#issuecomment-769030695.

We should hook it up to run in our CI. To run either on every PR (if it takes no more than 15 min) or in the nightly run (if it takes more).

The setup should be similar to what we do in [test/externalTests/](https://github.com/ethereum/solidity/tree/develop/test/externalTests) but there's no need to fork the repo in `solidity-external-test` - it should check out Buidler's repo directly.

These tests should be running also on the `breaking` branch.

### PR DESCRIPTION
**Pull request by cameel**
Fixes  #10854.
~Depends on https://github.com/nomiclabs/hardhat/pull/1962.~ Fixed.
~Depends on https://github.com/nomiclabs/hardhat/issues/1969.~ Fixed.
~Depends on https://github.com/nomiclabs/hardhat/pull/1974.~ Fixed.

### CODE REVIEW

**Comment by cameel** [comment]
There are two open hardhat PRs that need to be merged first:
- nomiclabs/hardhat#1962 fixes two failing tests on 0.8.5.
- nomiclabs/hardhat#1969 lets us use an absolute path to the soljson.js executable which removes the need for a workaround.

Also, I made it a PR check because it's pretty fast (runs in under a minute) but given that this depends on Hardhat's `master` branch, maybe it would be better to run it only nightly to avoid accidental breakage blocking our PRs?

**Comment by cameel** [comment]
All dependent issues have been resolved on Hardhat side and now `t_ems_ext_hardhat` succeeds so this is now ready for reviews.

File: .circleci/config.yml [hunk 1]
```diff
@@ -840,6 +840,32 @@ jobs:
...
+            yarn
+      - run:
+          name: Run hardhat-core test suite
+          command: |
+            HARDHAT_TESTS_SOLC_VERSION=$(scripts/get_version.sh)
+            export HARDHAT_TESTS_SOLC_VERSION
+
+            # NOTE: This is expected to work without running `yarn build` first.
+            cd hardhat/packages/hardhat-core
+            yarn test
```
Comment (hrkrshnn) [inline .circleci/config.yml:867]: The comment https://github.com/ethereum/solidity/issues/9237#issuecomment-769030695 mentioned running `scripts/run-tests-with-custom-solc.sh`. Would `yarn test` also work?

**Review by Marenz — APPROVED** [review]
LGTM

File: .circleci/config.yml (hunk 1 above)
Comment (Marenz) [reply .circleci/config.yml:867]: Are you asking if `yarn test` is enough or that it should be used?

Comment (cameel) [reply .circleci/config.yml:867]: It works. I took these commands from that script. The commands are simple enough that I think that it's better to run them directly. It also makes it possible to have separate steps in CI for more readable output.

Comment (hrkrshnn) [reply .circleci/config.yml:867]: Ah, just looked at the script. Nice.

**Review by hrkrshnn — APPROVED** [review]

**Pull request merged**